lucene/core/src/java/org/apache/lucene/index/SegmentInfo.java

```diff
@@ -326,7 +327,16 @@ public String getAttribute(String key) {
    * value.
    */
   public String putAttribute(String key, String value) {
-    return attributes.put(key, value);
+    HashMap<String, String> newMap = new HashMap<>(attributes);
+    String oldValue = newMap.put(key, value);
+    // we make a full copy of this to prevent concurrent modifications to this in the toString method
+    // this method is only called when a segment is written but the SegmentInfo might be exposed
+    // in running merges which can cause ConcurrentModificationExceptions if we modify / share
+    // the same instance. Technically that's an unsafe publication but IW design would require
+    // significant changes to prevent this. On the other hand, since we expose the map in getAttributes()
+    // it's a good design to make it unmodifiable anyway.
+    attributes = Collections.unmodifiableMap(newMap);
+    return oldValue;
```

Comment:
Can you fix the javadocs for this class to make it clear that every `putAttribute` call causes whole attributes `HashMap` to be cloned?